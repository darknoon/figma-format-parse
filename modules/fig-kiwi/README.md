## fig-kiwi parser

This module parses the figma file format:

- HTML file from pasteboard with (figma) (/figma) comments
- Data in .fig archives (starts with fig-kiwi)
- Includes typescript definitions derived from the [kiwi schema](https://github.com/evanw/kiwi) included with pasteboard and file data

Not included:

- Parsing Blob data in the returned Message (ie for vector networks etc)

Your use of this library is your responsibility, there is no warranty or support of any kind.

## Usage

```ts
import { readHTMLMessage } from "fig-kiwi";
const html = clipboardData.getData("text/html");
const { message, meta } = readHTMLMessage(html);
```

message data:

```json
{
  "type": "NODE_CHANGES",
  "sessionID": 0,
  "ackID": 0,
  "pasteID": 96190569,
  "pasteFileKey": "5z51beZU76UT5i92h3clNO",
  "pasteIsPartiallyOutsideEnclosingFrame": false,
  "pastePageId": {
    "sessionID": 0,
    "localID": 1
  }
  // ...
}
```

## Loading a .fig file in node

```ts
import { readFigFile, writeFigFile } from "fig-kiwi";
const fig: UInt8Array = readFileSync("vector.fig");
const { message, schema, preview } = readFigFile(fig);
// do something

// write out file
const mod = writeFigFile({ message, schema });
writeFileSync("modified.fig", mod);
```
