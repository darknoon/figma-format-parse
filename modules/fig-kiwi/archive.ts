export type Header = { prelude: string; version: number };

export default class FigmaArchiveParser {
  buffer: Uint8Array;
  data: DataView;
  offset = 0;

  constructor(buffer: Uint8Array) {
    this.buffer = buffer;
    this.data = new DataView(buffer.buffer);
  }

  readUint32(): number {
    // All little endian
    const n = this.data.getUint32(this.offset, true);
    this.offset += 4;
    return n;
  }

  read(bytes: number): Uint8Array {
    if (this.offset + bytes <= this.buffer.length) {
      const d = this.buffer.slice(this.offset, this.offset + bytes);
      this.offset += bytes;
      return d;
    } else {
      throw new Error(`read(${bytes}) is past end of data`);
    }
  }

  readHeader(): Header {
    const preludeExpected = "fig-kiwi";
    const preludeData = this.read(preludeExpected.length);
    const prelude = String.fromCharCode(...preludeData);
    if (prelude != preludeExpected) {
      throw new Error(`Unexpected prelude: "${prelude}"`);
    }
    const version = this.readUint32();
    return { prelude, version };
  }

  readData(size: number): Uint8Array {
    return this.read(size);
  }

  readAll(): { header: Header; files: Uint8Array[] } {
    const header = this.readHeader();
    const files: Uint8Array[] = [];
    while (this.offset + 4 < this.buffer.length) {
      const size = this.readUint32();
      const data = this.readData(size);
      files.push(data);
    }
    return { header, files };
  }
  static parseArchive(data: Uint8Array): {
    header: Header;
    files: Uint8Array[];
  } {
    const parser = new FigmaArchiveParser(data);
    return parser.readAll();
  }
}
