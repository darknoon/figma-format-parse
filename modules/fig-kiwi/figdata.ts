import { compileSchema, Schema } from "kiwi-schema";
import {
  Message,
  Schema as CompiledSchema,
  SparseMessage,
} from "./schema-defs";

export default class FigDataParser {
  data: Uint8Array;
  schema: Schema;
  constructor(data: Uint8Array, schema: Schema) {
    this.data = data;
    this.schema = schema;
  }

  parseMessage(): Message {
    const compiled = compileSchema(this.schema) as CompiledSchema;
    return compiled.decodeMessage(this.data);
  }

  parseSparseMessage(): SparseMessage {
    const compiled = compileSchema(this.schema) as CompiledSchema;
    return compiled.decodeSparseMessage(this.data);
  }

  /*
  tryParseAll(): { type: string; result: Object } | null {
    const compiled = compileSchema(this.schema);
    let decoded: any = null;
    let decodedName: string = "";
    const decoderNames = Object.keys(compiled).filter((d) =>
      d.startsWith("decode")
    );
    for (let name of decoderNames) {
      try {
        // Have to call with this syntax so `this` is bound to `compiled`
        decoded = compiled[name](this.data);
        decodedName = name;
        console.log("decoded as ", decodedName, decoded);
      } catch (err) {
        // console.error(err);
        // console.warn(name + "failed");
      }
    }
    if (decoded === null) {
      return null;
    }
    return { type: decodedName, result: decoded };
  }*/
}
