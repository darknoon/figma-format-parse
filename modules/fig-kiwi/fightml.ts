import { Buffer } from "buffer";

const metaStart = "<!--(figmeta)";
const metaEnd = "(/figmeta)-->";

const figmaStart = "<!--(figma)";
const figmaEnd = "(/figma)-->";

export interface FigmaMeta {
  fileKey: string;
  pasteID: number;
  dataType: "scene";
}
export interface ParsedFigmaHTML {
  meta: FigmaMeta;
  figma: Uint8Array;
}

function base64_decode_string(s: string): string {
  // TODO: don't require Buffer for this
  return Buffer.from(s, "base64").toString();
}

function base64_decode_data(s: string): Uint8Array {
  // TODO: don't require Buffer for this
  return Buffer.from(s, "base64");
}

// Or throw exception on error
export function parseHTMLString(html: string): ParsedFigmaHTML {
  const msi = html.indexOf(metaStart);
  const mei = html.indexOf(metaEnd);
  const fsi = html.indexOf(figmaStart);
  const fei = html.indexOf(figmaEnd);

  if (msi == -1) {
    throw new Error("Couldn't find start of figmeta");
  }

  const metaB64 = html.substring(msi + metaStart.length, mei);
  const figB64 = html.substring(fsi + figmaStart.length, fei);

  const metaStr = base64_decode_string(metaB64);
  return {
    meta: JSON.parse(metaStr),
    figma: base64_decode_data(figB64),
  };
}
