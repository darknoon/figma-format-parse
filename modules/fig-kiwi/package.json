{
    "name": "fig-kiwi",
    "version": "0.0.1",
    "description": "Reads/writes the figma file format for .fig files or text/html pasteboard data.",
    "dependencies": {
        "buffer": "^6.0.3",
        "kiwi-schema": "^0.4.7",
        "pako": "^2.0.4"
    },
    "devDependencies": {
        "@types/jest": "^28.1.4",
        "@types/pako": "^2.0.0",
        "esbuild": "^0.14.48",
        "esbuild-jest": "^0.5.0",
        "jest": "^28.1.2",
        "npm-dts": "^1.3.12",
        "typescript": "^4"
    },
    "files": [
        "dist/"
    ],
    "module": "dist/index.esm.js",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "scripts": {
        "build": "node build.js",
        "test:watch": "jest --watch",
        "test": "jest"
    }
}