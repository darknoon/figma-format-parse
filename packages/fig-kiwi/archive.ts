export type Header = { prelude: string; version: number };

const FIG_KIWI_PRELUDE = "fig-kiwi";
const FIG_KIWI_VERSION = 15;

export default class FigmaArchiveParser {
  buffer: Uint8Array;
  data: DataView;
  offset = 0;

  constructor(buffer: Uint8Array) {
    this.buffer = buffer;
    this.data = new DataView(buffer.buffer);
  }

  private readUint32(): number {
    // All little endian
    const n = this.data.getUint32(this.offset, true);
    this.offset += 4;
    return n;
  }

  private read(bytes: number): Uint8Array {
    if (this.offset + bytes <= this.buffer.length) {
      const d = this.buffer.slice(this.offset, this.offset + bytes);
      this.offset += bytes;
      return d;
    } else {
      throw new Error(`read(${bytes}) is past end of data`);
    }
  }

  private readHeader(): Header {
    const preludeData = this.read(FIG_KIWI_PRELUDE.length);
    // @ts-ignore todo: either downlevel-iteration or a type mismatch here
    const prelude = String.fromCharCode.apply(String, preludeData);
    if (prelude != FIG_KIWI_PRELUDE) {
      throw new Error(`Unexpected prelude: "${prelude}"`);
    }
    const version = this.readUint32();
    return { prelude, version };
  }

  private readData(size: number): Uint8Array {
    return this.read(size);
  }

  private readAll(): { header: Header; files: Uint8Array[] } {
    const header = this.readHeader();
    const files: Uint8Array[] = [];
    while (this.offset + 4 < this.buffer.length) {
      const size = this.readUint32();
      const data = this.readData(size);
      files.push(data);
    }
    return { header, files };
  }

  static parseArchive(data: Uint8Array): {
    header: Header;
    files: Uint8Array[];
  } {
    const parser = new FigmaArchiveParser(data);
    return parser.readAll();
  }
}

export class FigmaArchiveWriter {
  public header: Header;
  public files: Uint8Array[] = [];

  constructor() {
    this.header = { prelude: FIG_KIWI_PRELUDE, version: FIG_KIWI_VERSION };
  }

  /** 
   Add pre-compressed chunk to archive
  */

  write(): Uint8Array {
    // First find the total size, allocate a large enough buffer
    const headerSize = FIG_KIWI_PRELUDE.length + 4;
    const totalSize = this.files.reduce(
      (sz, f) => sz + 4 + f.byteLength,
      headerSize
    );
    const buffer = new Uint8Array(totalSize);
    const view = new DataView(buffer.buffer);
    const enc = new TextEncoder();
    // Write
    let offset = 0;
    offset = enc.encodeInto(FIG_KIWI_PRELUDE, buffer).written!;
    view.setUint32(offset, this.header.version, true);
    offset += 4;
    for (let file of this.files) {
      view.setUint32(offset, file.byteLength, true);
      offset += 4;
      buffer.set(file, offset);
      offset += file.byteLength;
    }
    return buffer;
  }
}
