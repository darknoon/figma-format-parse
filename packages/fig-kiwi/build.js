// @ts-check
const { build } = require("esbuild");
// @ts-ignore
const { dependencies, peerDependencies } = require("./package.json");
const { Generator, ELogLevel } = require("npm-dts");

// Following this guide https://souporserious.com/bundling-typescript-with-esbuild-for-npm/

new Generator(
  {
    entry: "index.ts",
    output: "dist/index.d.ts",
    logLevel: ELogLevel.warn,
  },
  true,
  true
)
  .generate()
  .then(() => console.log("Finished gen"));

const external = Object.keys(dependencies ?? {}).concat(
  Object.keys(peerDependencies ?? {})
);

const shared = {
  entryPoints: ["index.ts"],
  external,
  bundle: true,
};

build({
  ...shared,
  outfile: "dist/index.js",
});
build({ ...shared, format: "esm", outfile: "dist/index.esm.js" });
