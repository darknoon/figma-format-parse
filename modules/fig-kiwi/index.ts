import { composeHTMLString, FigmaMeta, parseHTMLString } from "./fightml";
import FigmaArchiveParser, { FigmaArchiveWriter, Header } from "./archive";
import {
  compileSchema,
  decodeBinarySchema,
  encodeBinarySchema,
  Schema,
} from "kiwi-schema";
import { Message, Schema as CompiledSchema } from "./schema-defs";
import { deflateRaw, inflateRaw } from "pako";
// Exports as the parsed default schema
import defaultSchema from "./schema";

export { type FigmaMeta };

interface ParsedFigma {
  header: Header;
  meta: FigmaMeta;
  schema: Schema;
  message: Message;
}

export function readHTMLMessage(html: string): ParsedFigma {
  const { figma, meta } = parseHTMLString(html);
  const { header, files } = FigmaArchiveParser.parseArchive(figma);
  const [schemaCompressed, dataCompressed] = files;
  const schema = decodeBinarySchema(inflateRaw(schemaCompressed));
  const compiledSchema = compileSchema(schema) as CompiledSchema;
  const message = compiledSchema.decodeMessage(inflateRaw(dataCompressed));
  return { header, meta, schema, message };
}

export function writeHTMLMessage(m: {
  meta: FigmaMeta;
  schema: Schema;
  message: Message;
}): string {
  const { meta, schema, message } = m;
  const encoder = new FigmaArchiveWriter();
  const binSchema = encodeBinarySchema(schema);
  const compiledSchema = compileSchema(schema) as CompiledSchema;
  encoder.files = [
    deflateRaw(binSchema),
    deflateRaw(compiledSchema.encodeMessage(m.message)),
  ];
  const data = encoder.write();
  return composeHTMLString({ meta, figma: data });
}

export function readFigFile(data: Uint8Array): {
  schema: Schema;
  header: Header;
  message: Message;
  preview: Uint8Array;
} {
  const { header, files } = FigmaArchiveParser.parseArchive(data);
  const [schemaFile, dataFile, preview] = files;
  const fileSchema = decodeBinarySchema(inflateRaw(schemaFile));
  const compiledSchema = compileSchema(fileSchema) as CompiledSchema;
  const message = compiledSchema.decodeMessage(inflateRaw(dataFile));
  return { message, schema: fileSchema, header, preview };
}

export function writeFigFile(settings: {
  schema?: Schema;
  header?: Header;
  message: Message;
  preview?: Uint8Array;
}): Uint8Array {
  const { schema = defaultSchema, message, preview } = settings;
  const compiledSchema = compileSchema(schema);
  const binSchema = encodeBinarySchema(schema);

  const writer = new FigmaArchiveWriter();
  writer.files = [
    deflateRaw(binSchema),
    deflateRaw(compiledSchema.encodeMessage(message)),
  ];
  if (preview) {
    writer.files.push(preview);
  }
  return writer.write();
}

export { FigmaArchiveParser };
