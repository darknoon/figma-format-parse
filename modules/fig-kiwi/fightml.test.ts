import { readFileSync, writeFileSync } from "fs";
import { parseHTMLString } from "./fightml";
import FigmaArchiveParser from "./archive";
import { compileSchema, decodeBinarySchema } from "kiwi-schema";
import { Schema as CompiledSchema, Message, SparseMessage } from "./fig-kiwi";
import { inflateRaw } from "pako";

test("parses components from html string", () => {
  const str = readFileSync(__dirname + "/data/c00000-circle-paste.html", {
    encoding: "utf-8",
  });

  const { figma, meta } = parseHTMLString(str);

  expect(meta).toEqual({
    fileKey: "whhd71XEXzYNauiHCeztfu",
    pasteID: 2130739331,
    dataType: "scene",
  });

  expect(figma.byteLength).toBeGreaterThan(1000);

  // Just to make debugging easier, extract the file
  writeFileSync(
    __dirname + "/data/c00000-circle-paste.fig",
    new Uint8Array(figma)
  );

  const parsed = FigmaArchiveParser.parseArchive(figma);
  expect(parsed.files.length).toBe(2);

  const [schemaFile, dataFile] = parsed.files;
  const fileSchema = decodeBinarySchema(inflateRaw(schemaFile));

  const cs = compileSchema(fileSchema) as CompiledSchema;

  const message: SparseMessage = cs.decodeSparseMessage(inflateRaw(dataFile));
  expect(message).toMatchSnapshot();
});
