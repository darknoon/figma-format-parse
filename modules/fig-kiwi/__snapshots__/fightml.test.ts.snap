// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses components from html string 1`] = `
Object {
  "ackID": 0,
  "isCut": false,
  "nodeChanges": Array [
    Object {
      "blendMode": "PASS_THROUGH",
      "guid": Object {
        "localID": 0,
        "sessionID": 0,
      },
      "mask": false,
      "maskIsOutline": false,
      "name": "Document",
      "opacity": 1,
      "phase": "CREATED",
      "type": "DOCUMENT",
      "visible": true,
    },
    Object {
      "backgroundColor": Object {
        "a": 1,
        "b": 0.9607843160629272,
        "g": 0.9607843160629272,
        "r": 0.9607843160629272,
      },
      "backgroundEnabled": true,
      "backgroundOpacity": 1,
      "blendMode": "PASS_THROUGH",
      "exportBackgroundDisabled": false,
      "guid": Object {
        "localID": 1,
        "sessionID": 0,
      },
      "mask": false,
      "maskIsOutline": false,
      "name": "Page 1",
      "opacity": 1,
      "parentIndex": Object {
        "guid": Object {
          "localID": 0,
          "sessionID": 0,
        },
        "position": "!",
      },
      "phase": "CREATED",
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "CANVAS",
      "visible": true,
    },
    Object {
      "arcData": Object {
        "endingAngle": 6.2831854820251465,
        "innerRadius": 0,
        "startingAngle": 0,
      },
      "blendMode": "PASS_THROUGH",
      "cornerRadius": 0,
      "cornerSmoothing": 0,
      "dashPattern": Array [],
      "fillPaints": Array [
        Object {
          "blendMode": "NORMAL",
          "color": Object {
            "a": 1,
            "b": 0,
            "g": 0,
            "r": 0.7541666626930237,
          },
          "opacity": 1,
          "type": "SOLID",
          "visible": true,
        },
      ],
      "guid": Object {
        "localID": 2,
        "sessionID": 2,
      },
      "handleMirroring": "NONE",
      "horizontalConstraint": "MIN",
      "locked": false,
      "mask": false,
      "maskIsOutline": false,
      "miterLimit": 4,
      "name": "Ellipse 1",
      "opacity": 1,
      "parentIndex": Object {
        "guid": Object {
          "localID": 1,
          "sessionID": 0,
        },
        "position": "!",
      },
      "phase": "CREATED",
      "proportionsConstrained": true,
      "size": Object {
        "x": 145,
        "y": 145,
      },
      "strokeAlign": "INSIDE",
      "strokeCap": "NONE",
      "strokeJoin": "MITER",
      "strokePaints": Array [],
      "strokeWeight": 1,
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "ELLIPSE",
      "verticalConstraint": "MIN",
      "visible": true,
    },
    Object {
      "backgroundColor": Object {
        "a": 1,
        "b": 0.9607843160629272,
        "g": 0.9607843160629272,
        "r": 0.9607843160629272,
      },
      "backgroundEnabled": true,
      "backgroundOpacity": 1,
      "blendMode": "PASS_THROUGH",
      "exportBackgroundDisabled": false,
      "guid": Object {
        "localID": 2,
        "sessionID": 20000002,
      },
      "internalOnly": true,
      "mask": false,
      "maskIsOutline": false,
      "name": "Internal Only Canvas",
      "opacity": 1,
      "parentIndex": Object {
        "guid": Object {
          "localID": 0,
          "sessionID": 0,
        },
        "position": "\\"",
      },
      "phase": "CREATED",
      "transform": Object {
        "m00": 1,
        "m01": 0,
        "m02": 0,
        "m10": 0,
        "m11": 1,
        "m12": 0,
      },
      "type": "CANVAS",
      "visible": false,
    },
  ],
  "pasteEditorType": "DESIGN",
  "pasteFileKey": "whhd71XEXzYNauiHCeztfu",
  "pasteID": 2130739331,
  "pasteIsPartiallyOutsideEnclosingFrame": false,
  "pastePageId": Object {
    "localID": 1,
    "sessionID": 0,
  },
  "sessionID": 0,
  "type": "NODE_CHANGES",
}
`;
